tests/test_qubit_device.py

```diff
@@ -0,0 +1,1121 @@
+
+        mock_qubit_device._wires_used = [1,2]
+        number_of_states = 2 ** len(mock_qubit_device._wires_used)
+
+        with monkeypatch.context() as m:
+            # Mock the auxiliary methods such that they return the expected values
+            m.setattr(QubitDevice, 'sample_basis_states', lambda self, wires, b: wires)
+            m.setattr(QubitDevice, 'states_to_binary', lambda a, b: (a, b))
+            mock_qubit_device.generate_samples()
+
+        assert mock_qubit_device._samples == (number_of_states, number_of_states)
+
+class TestSampleBasisStates:
+    """Test the sample_basis_states method"""
+
+    def test_sampling_with_correct_arguments(self, mock_qubit_device, monkeypatch):
+        """Tests that the sample_basis_states method samples with the correct arguments"""
+
+        shots = 1000
+
+        number_of_states = 4
+        mock_qubit_device.shots = shots
+        state_probs = [0.1, 0.2, 0.3, 0.4]
+
+        with monkeypatch.context() as m:
+            # Mock the numpy.random.choice method such that it returns the expected values
+            m.setattr("numpy.random.choice", lambda x, y, p: (x, y, p))
+            res = mock_qubit_device.sample_basis_states(number_of_states, state_probs)
+
+        assert np.array_equal(res[0], np.array([0, 1, 2, 3]))
+        assert res[1] == shots
+        assert res[2] == state_probs
+
+class TestStatesToBinary:
+    """Test the states_to_binary method"""
+
+    def test_correct_conversion_two_states(self, mock_qubit_device, monkeypatch):
+        """Tests that the sample_basis_states method converts samples to binary correctly"""
+        number_of_states = 2
+        basis_states = np.arange(number_of_states)
+        samples = np.random.choice(basis_states, mock_qubit_device.shots)
+
+        with monkeypatch.context() as m:
+            res = mock_qubit_device.states_to_binary(samples, number_of_states)
+
+        assert np.array_equal(res[:,0], samples)
+        assert np.array_equal(res[:,1], np.zeros(mock_qubit_device.shots))
+
+    @pytest.mark.parametrize("samples, binary_states",
+                            [
+                            (
+                              np.array([2, 3, 2, 0, 0]),
+                              np.array([[0, 1, 0, 0],
+                                        [1, 1, 0, 0],
+                                        [0, 1, 0, 0],
+                                        [0, 0, 0, 0],
+                                        [0, 0, 0, 0]],
+                             )
+                               ),
+                            (
+                            np.array([2, 3, 1, 3, 1]),
+                            np.array([[0, 1, 0, 0],
+                                       [1, 1, 0, 0],
+                                       [1, 0, 0, 0],
+                                       [1, 1, 0, 0],
+                                       [1, 0, 0, 0]])
+
```

Comment:
Maybe I am just getting a bit confused, but don't we use the convention in PennyLane that computational basis state 2 corresponds to |0010>, not |0100>?

Reply:
(As per our discussion)
It was totally confusing for me as well now that I saw it! :laughing: But so this one is used the following way in the code:
```python
self._samples[:, wires[0]]
```
Meaning that contrary to the mathematical representation, the first qubit will be the first columns, second qubit the second column, etc.

Thanks for the note, I'll add a comment regarding this!